Add documentation for smart contract upgradeability

limemloh opened the pull request:
## Purpose

Document smart contract upgradeability in the upcoming protocol update.

Closes https://github.com/Concordium/concordium-rust-smart-contracts/issues/178.

## Changes

- Add a guide of how to make a rust smart contract upgradeable.
- Add section about upgradeability in the chapter explaining smart contract instances.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

limemloh commented:
This should be release with the upcoming protocol update.

abizjak commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+- The new module does not contain a smart contract with a name matching the instance being upgraded.
+- The new module is a smart contract module version 0.
+
+Immutability
+------------
+
+It is still possible to implement an immutable smart contract, by simply `not` implementing any logic
+triggering the *upgrade* host function.
```
Can we make this even more prominent?

limemloh replied:
I have tried to improve this

abizjak commented on source/mainnet/smart-contracts/guides/upgradeable-contract.rst:
```diff
@@ -0,0 +1,67 @@
+       /// The new module reference.
+       module:  ModuleReference,
+       /// Migration function in the new module.
+       migrate: Option<(OwnedEntrypointName, Vec<u8>)>,
```
Don't you want to use the `OwnedParameter` type here?

abizjak replied:
Or is that not avialable in concordium-std?

limemloh replied:
it is available

Bargsteen commented on source/mainnet/smart-contracts/guides/upgradeable-contract.rst:
```diff
@@ -0,0 +1,67 @@
+The parameter for this endpoint takes the new module reference and optionally a name of an entrypoint with parameter to invoke in the upgraded smart contract instance.
+
+Providing the optional entrypoint can be used for triggering a migration function in the new module.
+This has the benefit of being in the same transaction as the upgrade itself, making the transaction reject reverting the upgrade if the migration failed.
```
```suggestion
This has the benefit of being in the same transaction as the upgrade itself, making the transaction revert the upgrade if the migration fails.
```

Bargsteen commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+module using the **upgrade** host function.
+The host function takes a reference to a deployed smart contract module to use for
+the upgraded instance and can only be called from a receive function.
+The function returns back whether the upgrade is successful or failed, allowing the instance
```
```suggestion
The host function returns whether the upgrade succeeded, allowing the instance
```

Bargsteen commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+The host function takes a reference to a deployed smart contract module to use for
+the upgraded instance and can only be called from a receive function.
+The function returns back whether the upgrade is successful or failed, allowing the instance
+to decide the next step. If the upgrade is succeful any new invocations of the
```
```suggestion
to decide the next step. If the upgrade is successful, any new invocations of the
```

Bargsteen commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+
+.. warning::
+
+   Make sure to proper authorized any endpoint triggering a smart contract upgrade as
+   it allows for changing the behavior of the smart contract.
```
Perhaps?
```suggestion
   Make sure to restrict access to any endpoint triggering a smart contract upgrade as
   it allows for changing the behavior of the smart contract.
```

Bargsteen commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+       escrow2[label="Escrow"]
+
+       House:n -> escrow;
+       Car:n -> escrow [style=dashed];
```
I think `dashed` is a bit hard to distinguish from the solid line. Consider using a different style.

limemloh replied:
Unfortunately, there are not that many options, but I have tried some other things to make it stand out a bit more.
Let me know what you think

Bargsteen replied:
Your changes made it much better. Great!

Bargsteen reviewed (commented):
Looks good. Just minor language issues ;)

limemloh commented on source/mainnet/smart-contracts/references/host-fns.rst:
```diff
@@ -373,6 +373,24 @@ Functions only accessible for smart contract receive functions.
+
+            ``2`` if failed because of module did not contain a smart contract with a name matching to one of this instance.
+
+            ``3`` if failed because of module being an unsupported smart contract version.
```
These have been updated

limemloh commented on source/mainnet/smart-contracts/guides/upgradeable-contract.rst:
```diff
@@ -0,0 +1,67 @@
+.. note::
+   A common way to workaround immutability, seen on other blockchains, is to have an additinal smart contract acting as a proxy, forwarding calls to the actual smart contract.
+   Upgrading is then updating the proxy to forward the calls to another smart contract.
+   Unfortunately, this will introduce an extra cost on every contract call and add an additional layer of complexity to the contract code.
```
I have added a bit more

limemloh commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,88 @@ If an operations fails, it returns an error, which the instance can choose to
+.. warning::
+
+   Make sure to proper authorized any endpoint triggering a smart contract upgrade as
+   it allows for changing the behavior of the smart contract.
```
I have tried to rephrase this

abizjak commented on source/mainnet/smart-contracts/general/contract-instances.rst:
```diff
@@ -168,3 +168,89 @@ If an operations fails, it returns an error, which the instance can choose to
+Immutability
+------------
+
+Only the smart contract itself can trigger an upgrade of its module, meaning that smart contracts
+are immutable, when it does not contain any code for triggering an upgrade.
```
Can we be a bit more precise here and also add a warning that just because they cannot be upgraded does not mean that the behaviour of a contract cannot change due to state changes, for example.
```suggestion
Only the smart contract itself can trigger an upgrade of its module, meaning that smart contracts
are immutable, when they do not contain any code for triggering an upgrade.
```

limemloh replied:
I have added a warning

abizjak commented on source/mainnet/smart-contracts/guides/upgradeable-contract.rst:
```diff
@@ -0,0 +1,67 @@
+
+This guide will show how to make a Rust smart contract upgradeable.
+
+Immutable smart contracts comes with the drawback that bugs cannot be fixed, new features and cost optimizations cannot be implemented.
```
```suggestion
Immutable smart contracts come with the drawback that bugs cannot be fixed, new features and cost optimizations cannot be implemented.
```

abizjak reviewed (approved):
A couple of minor clarifications, otherwise looks good.

chrmatt commented:
It would be nice to also include an example of how to perform an upgrade using concordium-client. This could be part of the on-chain guides and linked from the "Make a contract upgradeable" page.

chrmatt commented:
The documentation should mention which version of concordium-std is required for the upgradeability.

limemloh commented:
I have extended the snippet to derive a schema, included a json template for the parameter, added a note with the concordium-std version and added a link to guide for interacting with a contract using concordium-client using json.

chrmatt commented:
> I have extended the snippet to derive a schema, included a json template for the parameter, added a note with the concordium-std version and added a link to guide for interacting with a contract using concordium-client using json.

Is it decided that it will be concordium-std 5?
